crypto/keyring/keyring_test.go

```diff
@@ -29,69 +33,118 @@ func init() {
 	require.NoError(t, err)
-	require.Equal(t, "foo", info.GetName())
+	require.Equal(t, "foo", k.Name)
+}
+
+func requireEqualNames(records []*Record, n1, n2 string) bool {
+	for _, k := range records {
+		if k == nil {
+			continue
+		}
+
+		if k.Name != n2 && k.Name != n1 {
+			return false
+		}
+	}
+
+	return true
+}
+
+func getN2Index(records []*Record, n2 string) (j int) {
+	for i, k := range records {
+		if k == nil {
+			continue
+		}
+
+		if k.Name == n2 {
+			j = i
+			break
+		}
+	}
+
+	return j
 }
 
 func TestKeyManagementKeyRing(t *testing.T) {
-	kb, err := New("keybasename", "test", t.TempDir(), nil)
+	cdc := getCodec()
+	kb, err := New("keybasename", "test", t.TempDir(), nil, cdc)
 	require.NoError(t, err)
+	require.NotNil(t, cdc)
 
 	algo := hd.Secp256k1
 	n1, n2, n3 := "personal", "business", "other"
 
 	// Check empty state
-	l, err := kb.List()
-	require.Nil(t, err)
-	require.Empty(t, l)
+	records, err := kb.List()
+	require.NoError(t, err)
+	require.Len(t, records, 0)
 
 	_, _, err = kb.NewMnemonic(n1, English, sdk.FullFundraiserPath, DefaultBIP39Passphrase, notSupportedAlgo{})
 	require.Error(t, err, "ed25519 keys are currently not supported by keybase")
 
 	// create some keys
 	_, err = kb.Key(n1)
 	require.Error(t, err)
-	i, _, err := kb.NewMnemonic(n1, English, sdk.FullFundraiserPath, DefaultBIP39Passphrase, algo)
-
+	// save localKey with "n1`"
+	k, _, err := kb.NewMnemonic(n1, English, sdk.FullFundraiserPath, DefaultBIP39Passphrase, algo)
 	require.NoError(t, err)
-	require.Equal(t, n1, i.GetName())
-	_, _, err = kb.NewMnemonic(n2, English, sdk.FullFundraiserPath, DefaultBIP39Passphrase, algo)
+	require.Equal(t, n1, k.Name)
+
+	// save localKey with "n2"
+	k1, _, err := kb.NewMnemonic(n2, English, sdk.FullFundraiserPath, DefaultBIP39Passphrase, algo)
 	require.NoError(t, err)
+	require.Equal(t, n2, k1.Name)
 
-	// we can get these keys
-	i2, err := kb.Key(n2)
+	k2, err := kb.Key(n2)
 	require.NoError(t, err)
 	_, err = kb.Key(n3)
 	require.NotNil(t, err)
-	_, err = kb.KeyByAddress(accAddr(i2))
+	addr, err := k2.GetAddress()
+	require.NoError(t, err)
+	_, err = kb.KeyByAddress(addr)
 	require.NoError(t, err)
-	addr, err := sdk.AccAddressFromBech32("cosmos1yq8lgssgxlx9smjhes6ryjasmqmd3ts2559g0t")
+	addr, err = sdk.AccAddressFromBech32("cosmos1yq8lgssgxlx9smjhes6ryjasmqmd3ts2559g0t")
 	require.NoError(t, err)
 	_, err = kb.KeyByAddress(addr)
-	require.NotNil(t, err)
+	require.Error(t, err)
 
 	// list shows them in order
 	keyS, err := kb.List()
 	require.NoError(t, err)
-	require.Equal(t, 2, len(keyS))
+	require.Equal(t, 4, len(keyS))
 	// note these are in alphabetical order
-	require.Equal(t, n2, keyS[0].GetName())
-	require.Equal(t, n1, keyS[1].GetName())
-	require.Equal(t, i2.GetPubKey(), keyS[0].GetPubKey())
+	require.True(t, requireEqualNames(keyS, n1, n2))
+
+	key1, err := k2.GetPubKey()
+	require.NoError(t, err)
+	require.NotNil(t, key1)
+	i := getN2Index(keyS, n2)
```

Comment:
Are the key names not in alphabetical order anymore? Why do we need requireEqualNames and getN2Index?

Reply:
I tried to follow @amaurym recommendation and remove both `requireEqualNames` and `getIndices()`, but unfortunately it did not work out. The reason for that is the indices in `res`  inside `keyring.List()` are being assigned non deterministically while running `TestKeyManagementKeyRing` multiple times.